pkg/ipam/node.go

```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
+			if _, ok := n.ipsMarkedForRelease[ip]; !ok {
+				n.ipsMarkedForRelease[ip] = releaseTS
+			}
+		}
+	}
 	scopedLog := n.logger()
 
-	// Release excess addresses
-	if a.release != nil && len(a.release.IPsToRelease) > 0 {
+	// Attempt to release if :
+	// * IP is still marked for release in the current iteration's a.release.IPsToRelease
+	// * Marked for release for more than excess-ip-release-delay seconds
+	// * IP is not in use by any pod according to pod informer cache
+	var ipsToMark []string
+	var ipsToRelease []string
+
+	if n.ipsMarkedForRelease == nil || a.release == nil || len(a.release.IPsToRelease) == 0 {
+		// Resetting ipsMarkedForRelease if there are no IPs to release in this iteration
+		n.ipsMarkedForRelease = make(map[string]time.Time)
+	}
+	for markedIP, ts := range n.ipsMarkedForRelease {
+		// Determine which IPs are still marked for release.
+		stillMarkedForRelease := false
+		for _, ip := range a.release.IPsToRelease {
+			if markedIP == ip {
+				stillMarkedForRelease = true
+				break
+			}
+		}
+		if !stillMarkedForRelease {
+			// n.determineMaintenanceAction() only returns the IPs on the interface with maximum number of IPs that
+			// can be freed up. If the selected interface changes or if this IP is not excess anymore, remove entry
+			// from local map and cilium node status if n.ipReleaseStatus is already updated.
+			// from n.ipReleaseStatus
+			delete(n.ipsMarkedForRelease, markedIP)
+			delete(n.ipReleaseStatus, markedIP)
+			continue
+		}
+		// Check if the IP release waiting period elapsed
+		if ts.Add(time.Duration(operatorOption.Config.ExcessIPReleaseDelay) * time.Second).After(time.Now()) {
+			continue
+		}
+		// Handle IPs we've already heard back from agent.
+		if n.resource.Status.IPAM.ReleaseIPs != nil {
+			if status, ok := n.resource.Status.IPAM.ReleaseIPs[markedIP]; ok {
+				switch status {
+				case ipamOption.IPAMReadyForRelease:
+					ipsToRelease = append(ipsToRelease, markedIP)
+				case ipamOption.IPAMDoNotRelease:
+					scopedLog.WithFields(logrus.Fields{logfields.IPAddr: markedIP}).Debug("IP release request denied from agent")
+					delete(n.ipsMarkedForRelease, markedIP)
+					delete(n.ipReleaseStatus, markedIP)
+				}
+				continue
+			}
+		}
+		ipsToMark = append(ipsToMark, markedIP)
+	}
+	// Update cilium node CRD with IPs that need to be marked for release
+	for _, ip := range ipsToMark {
+		scopedLog.WithFields(logrus.Fields{"ip": ip}).Debug("Marking IP for release")
+		n.ipReleaseStatus[ip] = ipamOption.IPAMMarkForRelease
+	}
+
+	if len(ipsToRelease) > 0 {
+		a.release.IPsToRelease = ipsToRelease
+		scopedLog = scopedLog.WithFields(logrus.Fields{
+			"available":         n.stats.AvailableIPs,
+			"used":              n.stats.UsedIPs,
+			"excess":            n.stats.ExcessIPs,
+			"excessIps":         a.release.IPsToRelease,
+			"releasing":         ipsToRelease,
+			"selectedInterface": a.release.InterfaceID,
+			"selectedPoolID":    a.release.PoolID,
```

Comment:
Ideally, since these are a part of an INFO level log, they should be inside the logfields package. However, there are other parts of the code that also suffer the same thing, so let's just followup on this later, doesn't have to be part of this PR.